File: src/bin/0232-implement-queue-using-stacks.py
# @leetup=info id=232 lang=python3 slug=implement-queue-using-stacks



from typing import Optional

# @leetup=code
class MyQueue:
    def __init__(self):
        self.s1 = []
        self.s2 = []

    def push(self, x: int) -> None:
        self.s1.append(x)

    def pop(self) -> int:
        self.move_s1_to_s2_if_s2_is_empty()
        return self.s2.pop()

    def peek(self) -> int:
        self.move_s1_to_s2_if_s2_is_empty()
        return self.s2[-1]

    def move_s1_to_s2_if_s2_is_empty(self):
        if not self.s2:
            # s1: [1, 2, 3]
            # s2: []
            while self.s1:
                self.s2.append(self.s1.pop())
            # s1: []
            # s2: [3, 2, 1]

    def empty(self) -> bool:
        return (not self.s1) and (not self.s2)


# @leetup=code

File: src/bin/0232-implement-queue-using-stacks.rs
// @leetup=info id=232 lang=rust slug=implement-queue-using-stacks

// @leetup=code

#[allow(unused_imports)]
use std::collections::*;

struct MyQueue {
    s1: Vec<i32>,
    s2: Vec<i32>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyQueue {
    fn new() -> Self {
        MyQueue {
            s1: Vec::new(),
            s2: Vec::new(),
        }
    }

    fn push(&mut self, x: i32) {
        self.s1.push(x);
    }

    fn pop(&mut self) -> i32 {
        self.move_s1_to_s2_if_s2_is_empty();
        self.s2.pop().unwrap()
    }

    fn move_s1_to_s2_if_s2_is_empty(&mut self) {
        if self.s2.is_empty() {
            while let Some(n) = self.s1.pop() {
                self.s2.push(n);
            }
        }
    }

    fn peek(&mut self) -> i32 {
        self.move_s1_to_s2_if_s2_is_empty();
        *self.s2.last().unwrap()
    }

    pub fn empty(&self) -> bool {
        self.s1.is_empty() && self.s2.is_empty()
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * let obj = MyQueue::new();
 * obj.push(x);
 * let ret_2: i32 = obj.pop();
 * let ret_3: i32 = obj.peek();
 * let ret_4: bool = obj.empty();
 */
// @leetup=code

struct Solution;

fn main() {
    let _ = MyQueue::new();
}
